dask/layers.py

```diff
@@ -1,14 +1,17 @@
 import operator
 from collections import defaultdict
+from itertools import product
 from typing import List, Optional, Tuple
 
 import tlz as toolz
+from tlz.curried import map
 
+from .array.core import concatenate3, reshapelist
```

Comment:
This looks like it's causing circular import problems.

Reply:
I kinda think this should be in a separate file. I think `layers.py` will get too crowded if we try to put everything in there; maybe it should contain only core layers that are reused across data types (and I don't think there are any currently)? Curious what @rjzamora thinks. But I'd vote for moving all the dataframe stuff to `dask/dataframe/layers.py` or something, and have a `dask/array/layers.py`. Or potentially even a layers subdirectory, depending on how lengthy these get.

As for the "needs NumPy", I think there's just no way we'll be able to generate graphs for all these NumPy operations without importing NumPy to put some NumPy functions in the graph. Now, if we don't want to import NumPy on the scheduler, we could implement some sort of serialization trick, where we could stick named references to NumPy functions in the graph on the scheduler side, which would deserialize as the actual NumPy functions on the worker end. So as long as we don't have to _call_ NumPy functions in order to generate the layers, then I'm not too worried about making NumPy optional on the scheduler in the future.

Reply:
We have this kind of numpy call happening in the concatenate3 function, am I wrong in thinking that might cause a problem? https://github.com/dask/dask/blob/21aaf44d4d25bdba05951b85f3f2d943b823e82d/dask/array/core.py#L4812

Reply:
I think that's all fine if we use something akin to the serialization trick I mentioned. Looks like we already have a pattern for this actually: the `CallableLazyImport` in `ShuffleLayer`s: https://github.com/dask/dask/blob/23c93d72c8fbd194b44614c29dd584ff93fc1eef/dask/layers.py#L327-L337

And then we stick that `concat_func` into the graph: https://github.com/dask/dask/blob/23c93d72c8fbd194b44614c29dd584ff93fc1eef/dask/layers.py#L345-L349

So long as we don't need to call any NumPy functions at graph _construction_ time—and we'd only be importing them  to insert them in the graph—we should be fine. In this case, I think you could use a `CallableLazyImport` for `concatenate3`, for instance.

Reply:
Yes, this is exactly the type of scenario for which `CallableLazyImport` was created. For something like `reshapelist`, which is python-only, that might be able to be moved to `dask.utils`

As for creating `dask.array.layers` and `dask.dataframe.layers` modules, I share your concern about `layers.py` getting overstuffed, but I'm not sure that there is a way to make those submodules without importing `pandas`/`numpy`. It really is a challenging constraint!